### ISSUE TITLE
Crash with iOS release mode on some devices.

### ISSUE THREAD

**Issue by Wackymax** [flutter/flutter#144594]
### Steps to reproduce

Unable to reproduce and seems to happen only on some devices with Impeller.

Mostly seen on iPhone 14 Pro Max running iOS 17.3.1

### Expected results

Framework should not crash when rendering

### Actual results

Crashed: io.flutter.1.raster
0  Flutter                        0x527538 (Missing UUID 4c4c447355553144a149a4a73b6eae0b)
1  Flutter                        0x51fbe8 (Missing UUID 4c4c447355553144a149a4a73b6eae0b)
2  Flutter                        0x343d48 (Missing UUID 4c4c447355553144a149a4a73b6eae0b)
3  Flutter                        0x524a58 (Missing UUID 4c4c447355553144a149a4a73b6eae0b)
4  Flutter                        0x343c48 (Missing UUID 4c4c447355553144a149a4a73b6eae0b)
5  Flutter                        0x5cac78 InternalFlutterGpu_Texture_AsImage + 47768
6  Flutter                        0x4b767c (Missing UUID 4c4c447355553144a149a4a73b6eae0b)
7  Flutter                        0x3f33b0 (Missing UUID 4c4c447355553144a149a4a73b6eae0b)
8  Flutter                        0x3f389c (Missing UUID 4c4c447355553144a149a4a73b6eae0b)
9  Flutter                        0x2ea89c (Missing UUID 4c4c447355553144a149a4a73b6eae0b)
10 Flutter                        0x3f2ae8 (Missing UUID 4c4c447355553144a149a4a73b6eae0b)
11 Flutter                        0x3f46ec (Missing UUID 4c4c447355553144a149a4a73b6eae0b)
12 Flutter                        0x3f3f38 (Missing UUID 4c4c447355553144a149a4a73b6eae0b)
13 Flutter                        0x40c3ec (Missing UUID 4c4c447355553144a149a4a73b6eae0b)
14 Flutter                        0x2e8f08 (Missing UUID 4c4c447355553144a149a4a73b6eae0b)
15 Flutter                        0x2ec6bc (Missing UUID 4c4c447355553144a149a4a73b6eae0b)
16 CoreFoundation                 0xad3c8 <redacted> + 32
17 CoreFoundation                 0xad070 <redacted> + 1004
18 CoreFoundation                 0x36c04 <redacted> + 288
19 CoreFoundation                 0x33c1c <redacted> + 1856
20 CoreFoundation                 0x333f8 CFRunLoopRunSpecific + 608
21 Flutter                        0x2ec7a8 (Missing UUID 4c4c447355553144a149a4a73b6eae0b)
22 Flutter                        0x2ebe20 (Missing UUID 4c4c447355553144a149a4a73b6eae0b)
23 libsystem_pthread.dylib        0x24d4 _pthread_start + 136
24 libsystem_pthread.dylib        0x1a10 thread_start + 8

### Code sample

<details open><summary>Code sample</summary>

N/A

</details>

### Screenshots or Video

<details open>
<summary>Screenshots / Video demonstration</summary>

N/A

</details>

### Logs

<details open><summary>Logs</summary>

N/A

</details>

### Flutter Doctor output

<details open><summary>Doctor output</summary>

```console
[✓] Flutter (Channel stable, 3.19.2, on macOS 14.2.1 23C71 darwin-arm64, locale en-ZA)
    • Flutter version 3.19.2 on channel stable at /Users/hennie/Tools/Flutter/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 7482962148 (6 days ago), 2024-02-27 16:51:22 -0500
    • Engine revision 04817c99c9
    • Dart version 3.3.0
    • DevTools version 2.31.1

[✓] Android toolchain - develop for Android devices (Android SDK version 34.0.0)
    • Android SDK at /Users/hennie/Library/Android/sdk
    • Platform android-34, build-tools 34.0.0
    • ANDROID_SDK_ROOT = /Users/hennie/Library/Android/sdk
    • Java binary at: /Users/hennie/Applications/Android Studio.app/Contents/jbr/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 17.0.9+0-17.0.9b1087.7-11185874)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 15.2)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Build 15C500b
    • CocoaPods version 1.14.3

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2023.2)
    • Android Studio at /Users/hennie/Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 17.0.9+0-17.0.9b1087.7-11185874)

[✓] IntelliJ IDEA Ultimate Edition (version 2023.3.4)
    • IntelliJ at /Users/hennie/Applications/IntelliJ IDEA Ultimate.app
    • Flutter plugin version 78.1.1
    • Dart plugin version 233.13763.5

[✓] VS Code (version 1.86.0)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.80.0

[✓] Connected device (3 available)
    • Hennie iPhone (mobile) • 00008110-001818AA02A2801E • ios            • iOS 17.3.1 21D61
    • macOS (desktop)        • macos                     • darwin-arm64   • macOS 14.2.1 23C71 darwin-arm64
    • Chrome (web)           • chrome                    • web-javascript • Google Chrome 122.0.6261.94
    ! Error: Browsing on the local area network for Hennie Brink’s iPad. Ensure the device is unlocked and attached with a cable or associated with the same local area network as this Mac.
      The device must be opted into Developer Mode to connect wirelessly. (code -27)

[✓] Network resources
    • All expected network resources are available.

• No issues found!

```

</details>

**Comment by Wackymax** [comment]
@flutter-symbolizer-bot engine#04817c99c9 release arm64 https://github.com/flutter/flutter/issues/144594#issue-2168487679

**Comment by huycozy** [comment]
@flutter-symbolizer-bot https://github.com/flutter/flutter/issues/144594#issue-2168487679 flutter#v3.19.2 force ios release arm64

**Comment by huycozy** [comment]
It seems the bot can't symbolicate the crash. Does this happen locally on your machine or in production? Please share a full crash log if there is any.

For `InternalFlutterGpu_Texture_AsImage` trace, it looks similar to https://github.com/flutter/flutter/issues/140710.

**Comment by Wackymax** [comment]
> It seems the bot can't symbolicate the crash. Does this happen locally on your machine or in production? Please share a full crash log if there is any.
> 
> For `InternalFlutterGpu_Texture_AsImage` trace, it looks similar to #140710.

It's happening in production. This is the crashlog from Crashlytics. 
[com.flygaggle.flutterapp_issue_a264647685916a3d6013ee5ee56a72c6_crash_session_1d90b6048e1b4a07bc28853651c13345_DNE_8_v2_stacktrace.txt](https://github.com/flutter/flutter/files/14493364/com.flygaggle.flutterapp_issue_a264647685916a3d6013ee5ee56a72c6_crash_session_1d90b6048e1b4a07bc28853651c13345_DNE_8_v2_stacktrace.txt)

**Comment by huycozy** [comment]
Thanks for providing it. I see there are a lot of `<redacted>` traces in the stack. This looks similar to https://github.com/flutter/flutter/issues/140710. That being said, without unredacted stack traces or reproduction cases, we are unsure how to investigate this. If you can narrow down the issue and share a reproduction, it would be helpful.

**Comment by Wackymax** [comment]
Unfortunately it's only happening to a small subset of my users so I don't think it will be easy to identify the cause. I'll switch back to Skia for now. It seems like Impeller still has lots of bugs and performance problems, especially in my app that uses a lot of custom paint functionality. I really hope Skia is here to stay for a long time since it seems to be the better engine.

**Comment by matanlurey** [comment]
I'm confused how `InternalFlutterGpu_Texture_AsImage` is even making it into production apps:

https://github.com/search?q=repo%3Aflutter%2Fengine%20%22InternalFlutterGpu_Texture_AsImage%22&type=code

@bdero @jonahwilliams?

**Comment by bdero** [comment]
I don't think has anything to do with `InternalFlutterGpu_Texture_AsImage`. The address being shown in the trace is `InternalFlutterGpu_Texture_AsImage + [really big offset]`. So I think it's just printing the address relative to the closest previous exported/known symbol.

**Comment by matanlurey** [comment]
Ack, interesting...

**Comment by Wackymax** [comment]
That's understandable. Might not even be Impeller related in that case and it might be a 3rd party lib? This is all I can get from Crashlytics. Unless there is another process I can follow to get a better stack trace?

**Comment by jonahwilliams** [comment]
Crashlytics should have documentation on uploading dsym files: https://firebase.google.com/docs/crashlytics/get-deobfuscated-reports?platform=ios

IIRC if you do this the traces will be automatically symbolized.

**Comment by Wackymax** [comment]
I do have that setup but it looks it's only working for my code?On 06 Mar 2024, at 20:18, Jonah Williams ***@***.***> wrote:﻿
Crashlytics should have documentation on uploading dsym files: https://firebase.google.com/docs/crashlytics/get-deobfuscated-reports?platform=ios
IIRC if you do this the traces will be automatically symbolized.

**Comment by jonahwilliams** [comment]
I can't really walk through why symbol uploads may or may not have worked. i would ensure that you also uploaded the flutter engine dsym.

**Comment by Wackymax** [comment]
What does the Flutter engine dsym look like, the name of it? Looks like the build process only produces dsym files for my code base and all the packages I included.

**Comment by Wackymax** [comment]
@flutter-symbolizer-bot internal arm64 engine#04817c99c9 https://github.com/flutter/flutter/issues/144594#issue-2168487679

**Comment by jonahwilliams** [comment]
oh right the bot.

@flutter-symbolizer-bot force ios arm64 engine#04817c99c9 https://github.com/flutter/flutter/issues/144594#issue-2168487679

**Comment by jonahwilliams** [comment]
@Wackymax all I can find is https://stackoverflow.com/questions/55709939/how-to-upload-dsyms-files-which-developed-with-flutter

**Comment by Wackymax** [comment]
> @Wackymax all I can find is https://stackoverflow.com/questions/55709939/how-to-upload-dsyms-files-which-developed-with-flutter

Yeah this is happening already. The symbols for my code and all the packages are uploaded to Crashlytics. But it doens't include the Flutter engine code. I think it's because the Flutter engine is precompiled

**Comment by Wackymax** [comment]
Anyone know where I can get the dSym files for engine 04817c99c9? That way I can maybe upload it to Crashlytics

**Comment by Wackymax** [comment]
Crashed: io.flutter.1.raster
0  Flutter                        0x527538 (Missing UUID 4c4c447355553144a149a4a73b6eae0b)
1  Flutter                        0x51fbe8 (Missing UUID 4c4c447355553144a149a4a73b6eae0b)
2  Flutter                        0x343d48 (Missing UUID 4c4c447355553144a149a4a73b6eae0b)
3  Flutter                        0x344674 (Missing UUID 4c4c447355553144a149a4a73b6eae0b)
4  Flutter                        0x524a58 (Missing UUID 4c4c447355553144a149a4a73b6eae0b)
5  Flutter                        0x343c48 (Missing UUID 4c4c447355553144a149a4a73b6eae0b)
6  Flutter                        0x5cac78 InternalFlutterGpu_Texture_AsImage + 47768
7  Flutter                        0x4b767c (Missing UUID 4c4c447355553144a149a4a73b6eae0b)
8  Flutter                        0x3f33b0 (Missing UUID 4c4c447355553144a149a4a73b6eae0b)
9  Flutter                        0x3f389c (Missing UUID 4c4c447355553144a149a4a73b6eae0b)
10 Flutter                        0x2ea89c (Missing UUID 4c4c447355553144a149a4a73b6eae0b)
11 Flutter                        0x3f2ae8 (Missing UUID 4c4c447355553144a149a4a73b6eae0b)
12 Flutter                        0x3f46ec (Missing UUID 4c4c447355553144a149a4a73b6eae0b)
13 Flutter                        0x3f3f38 (Missing UUID 4c4c447355553144a149a4a73b6eae0b)
14 Flutter                        0x40c3ec (Missing UUID 4c4c447355553144a149a4a73b6eae0b)
15 Flutter                        0x2e8f08 (Missing UUID 4c4c447355553144a149a4a73b6eae0b)
16 Flutter                        0x2ec6bc (Missing UUID 4c4c447355553144a149a4a73b6eae0b)
17 CoreFoundation                 0xad3c8 <redacted> + 32
18 CoreFoundation                 0xad070 <redacted> + 1004
19 CoreFoundation                 0x36c04 <redacted> + 288
20 CoreFoundation                 0x33c1c <redacted> + 1856
21 CoreFoundation                 0x333f8 CFRunLoopRunSpecific + 608
22 Flutter                        0x2ec7a8 (Missing UUID 4c4c447355553144a149a4a73b6eae0b)
23 Flutter                        0x2ebe20 (Missing UUID 4c4c447355553144a149a4a73b6eae0b)
24 libsystem_pthread.dylib        0x24d4 _pthread_start + 136
25 libsystem_pthread.dylib        0x1a10 thread_start + 8

**Comment by jonahwilliams** [comment]
Should be here: https://github.com/flutter/flutter/wiki/Crashes#ios

**Comment by Wackymax** [comment]
@flutter-symbolizer-bot internal arm64 engine#04817c99c9 https://github.com/flutter/flutter/issues/144594#issuecomment-1981673840

**Comment by Wackymax** [comment]
Trying but I am not finding anything there. Based on the information in git for v3.19.2 the engine number should be as stated here: https://github.com/flutter/flutter/blob/3.19.2/bin/internal/engine.version

But the storage is empty: https://console.cloud.google.com/storage/browser/flutter_infra/flutter/04817c99c9fd4956f27505204f7e344335810aed

**Comment by Wackymax** [comment]
> Should be here: https://github.com/flutter/flutter/wiki/Crashes#ios

Thank you, found it. Uploading to Crashlytics now and will see what happens

**Comment by Wackymax** [comment]
I uploaded the engine dsym files to Crashlytics but the trace still looks the same. It doesn't look like we will get to the bottom of it with the traces that I have available. I did a new release switching to Skia and everything is working again for the people who experienced the crash so it does seem to be Impeller related. Hopefully someone else also runs into the problem and can get some better traces.
